crates/store/re_server/src/store/table.rs

```diff
@@ -78,4 +93,106 @@ impl Table {
+        }
+    }
+
+    pub fn provider(&self) -> Arc<dyn TableProvider> {
+        match &self.table {
+            TableType::DataFusionTable(t) => Arc::clone(t),
+            #[cfg(feature = "lance")]
+            TableType::LanceDataset(dataset) => Arc::new(LanceTableProvider::new(
+                Arc::new(dataset.as_ref().clone()),
+                false,
+                false,
+            )),
+        }
+    }
+
+    async fn write_table_provider(
+        &self,
+        rb: RecordBatch,
+        insert_op: InsertOp,
+    ) -> Result<(), DataFusionError> {
+        let schema = rb.schema();
+        let TableType::DataFusionTable(provider) = &self.table else {
+            return exec_err!("Expected DataFusion Table Provider");
+        };
+
+        let input = MemorySourceConfig::try_new_from_batches(schema, vec![rb])?;
+        let session = SessionStateBuilder::default().build();
+        let result = provider.insert_into(&session, input, insert_op).await?;
+        let mut output = result.execute(0, session.task_ctx())?;
+
+        while let Some(r) = output.next().await {
+            let _ = r?;
+        }
+        Ok(())
+    }
+
+    #[cfg(feature = "lance")]
+    async fn write_table_lance_dataset(
+        &mut self,
+        rb: RecordBatch,
+        insert_op: InsertOp,
+    ) -> Result<(), DataFusionError> {
+        let schema = rb.schema();
+        let mut params = WriteParams::default();
+
+        let TableType::LanceDataset(dataset) = &mut self.table else {
+            return exec_err!("Expected Lance Dataset");
+        };
+
+        let reader = RecordBatchIterator::new(vec![Ok(rb)], schema);
+
+        match insert_op {
+            InsertOp::Append => {
+                params.mode = WriteMode::Append;
+
+                dataset
+                    .as_ref()
+                    .clone()
+                    .append(reader, Some(params))
+                    .await
+                    .map_err(|err| DataFusionError::External(err.into()))?;
+            }
+            InsertOp::Replace => {
+                exec_err!("Invalid insert operation. Only append and overwrite are supported.")?;
+            }
+            InsertOp::Overwrite => {
+                params.mode = WriteMode::Overwrite;
+
+                let _ =
+                    LanceDataset::write(reader, Arc::new(dataset.as_ref().clone()), Some(params))
+                        .await
+                        .map_err(|err| DataFusionError::External(err.into()))?;
+            }
+        }
+
+        let updated_table = Arc::new(
+            lance::Dataset::open(dataset.uri())
+                .await
+                .map_err(|err| std::io::Error::new(std::io::ErrorKind::InvalidInput, err))?,
+        );
+
+        self.table = TableType::LanceDataset(updated_table);
+        Ok(())
+    }
+
+    pub async fn write_table(
+        &mut self,
+        rb: RecordBatch,
+        insert_op: InsertOp,
+    ) -> Result<(), DataFusionError> {
```

Comment:
For such `pub` methods used by the actual server hander (`rerun_cloud.rs`), I'm trying to use the `crate::Error` type. Motivation: this type tries to inject a high-level task semantic ("I'm trying to write a table") in top of whatever low-level error you get ("obscure metadata does not comply with moon phase").

I think that here this approach applies.

Note that for the reason above, I try to avoid `#[from]` stuff in `crate::Error`, relying on enum variant that explicitly capture the task at hand.